tools/debugging/matrix/check_federation_status.py

```diff
@@ -0,0 +1,116 @@
+from raiden.utils.signer import LocalSigner
+
+log = get_logger(__name__)
+
+
+@dataclass
+class FederationError:
+    source_server: str
+    target_sever: Optional[str]
+    reason: str
+
+
+class FederationChecker:
+    def __init__(self, known_servers_list_url: str) -> None:
+        self._known_servers_list_url = known_servers_list_url
+        self._seed: str = "".join(random.choice(string.ascii_letters) for _ in range(10))
+        self._clients: Dict[str, GMatrixClient] = {}
+        self._errors: Dict[str, List[Tuple[str, Exception]]] = defaultdict(list)
+
+    def check(self) -> List[FederationError]:
+        clients, errors = self._make_clients()
+        errors.extend(self._run_check(clients))
+        return errors
+
+    def _make_clients(self) -> Tuple[Dict[str, GMatrixClient], List[FederationError]]:
+        def _dummy_callback(_: MatrixSyncMessages) -> bool:
+            return True
+
+        errors: List[FederationError] = []
+        clients: Dict[str, GMatrixClient] = {}
+
+        known_servers = get_matrix_servers(
+            self._known_servers_list_url, ServerListType.ALL_SERVERS
+        )
+
+        for current_server in known_servers:
+            signer = LocalSigner(
+                hashlib.sha256(f"pk-{self._seed}-{current_server}".encode()).digest()
+            )
+            try:
+                clients[current_server] = client = make_client(_dummy_callback, [current_server])
+                login(client=client, signer=signer, device_id=DeviceIDs.RAIDEN)
+            except (TransportError, MatrixRequestError) as ex:
+                error = FederationError(
+                    source_server=current_server,
+                    target_sever=None,
+                    reason=f"Could not connect to {current_server}: {ex}",
+                )
+                log.error("Error connecting", error=error)
+                errors.append(error)
+        return clients, errors
+
+    @staticmethod
+    def _run_check(clients: Dict[str, GMatrixClient]) -> List[FederationError]:
+        results: List[FederationError] = []
+        for (source_server, source_client), (target_sever, target_client) in permutations(
+            clients.items(), 2
+        ):
+            log.info("Checking", source_server=source_server, target_sever=target_sever)
+            try:
+                source_client.get_user(target_client.user_id).get_display_name()
+            except Exception as ex:
+                error = FederationError(
+                    source_server=source_server,
+                    target_sever=target_sever,
+                    reason=f"Couldn't fetch profile: {ex}",
+                )
+                log.error("Error fetching profile", error=error, exception=str(ex))
+                results.append(error)
+        return results
+
+
+@click.command()
+@click.option("-l", "--known-servers-list-url", required=True)
+@click.option("--seed", required=True, help="Seed for private key")
```

Comment:
Is there any benefit controlling the seed? Do we use the same pk-seeding scheme somewhere else?

Reply:
Yes for one thing it prevents spamming the servers with new users every time the the tool is run (in case we were to just use random private keys) while also allowing to deterministically (re-)use the same users without having to pre-generate a bunch of private keys.

We use the same in other tools in the RSB.